### PR TITLE
Use feature markers during terminal-by-default handoff

### PR DESCRIPTION
**Pull request by lhecker**
If we want to make Windows Terminal the default terminal under Windows,
we'll have to make conhost "handoff" incoming connections by default.
But this poses a problem: How can the seldomly updated conhost know
whether the routinely updated Windows Terminal version is actually willing
to accept such handoffs by default (it might be unwilling due to bugs, etc.)?

This commit solves the issue by introducing:
* A marker interface (`IDefaultTerminalMarker`): If it exists,
  Windows Terminal indicates its willingness to accept the handoff.
* Turning the all-0 GUID from being synonymous for conhost,
  to being synonymous for "Let Windows decide". Without this we wouldn't
  be able to differentiate between users who consciously chose conhost
  as their default terminal, vs. users who want the standard behavior.

## Validation Steps Performed
Testing fallback behavior:
* Install "Terminal" 1.13
* Delete the 2 keys below `HKCU\Console\%%Startup`
* Enable `Feature_AttemptHandoff` in `features.xml`
  Return `true` from `DefaultApp::CheckShouldTerminalBeDefault`
* Replace `conhost.exe` and `console.dll` with `sfpcopy` after building
* Launching `cmd.exe` launches as a conhost window ✅
  (because "Terminal" 1.13 lacks the marker interface)
* Open properties page in `conhost.exe`
  "Let Windows decide" is select by default ✅
* Changing the selection writes the new value ✅

Testing the new behavior:
* Delete the 2 keys below `HKCU\Console\%%Startup`
* Enable `Feature_AttemptHandoff` in `features.xml`
  Return `true` from `DefaultApp::CheckShouldTerminalBeDefault`
* Use `CLSID_WindowsTerminalConsoleDev` and `CLSID_WindowsTerminalTerminalDev`
  for the initialization of `TerminalDelegationPair`
* Replace `conhost.exe` and `console.dll` with `sfpcopy` after building
* Deploy the "Terminal Dev" package
* Launching `cmd.exe` launches "Terminal Dev" ✅
  (because "Terminal Dev" has the marker interface)
* Open the settings tab
  "Let Windows decide" is select by default ✅
* Changing the selection and saving writes the new value ✅

### CODE REVIEW

File: src/host/srvinit.cpp [hunk 1]
```diff
@@ -439,11 +441,7 @@ try
 
     if (!g.handoffTerminalClsid)
     {
-        TraceLoggingWrite(g_hConhostV2EventTraceProvider,
-                          "SrvInit_ReceiveHandoff_NoTerminal",
-                          TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE),
-                          TraceLoggingKeyword(TIL_KEYWORD_TRACE));
-        return E_NOT_SET;
+        g.handoffTerminalClsid = CLSID_SystemDelegationTerminal;
```
Comment (lhecker) [inline src/host/srvinit.cpp:444]: I think we should introduce something like a `IConsoleHandoff2` interface as a follow-up that provides the correct `handoffTerminalClsid` directly as a handoff-parameter.
It might make sense to wait a little bit for that so that we can see whether any other parameters should be added...

File: src/server/IoDispatchers.cpp [hunk 2]
```diff
@@ -178,27 +178,15 @@ static bool _shouldAttemptHandoff(const Globals& globals,
...
-    {
-        return false;
-    }
-
-    // This console was started with a command line argument to
-    // specifically block handoff to another console. We presume
-    // this was for good reason (compatibility) and give up here.
-    if (globals.launchArgs.GetForceNoHandoff())
+    // If we do not have a registered handoff, do not attempt.
+    if (!globals.handoffConsoleClsid)
```
Comment (lhecker) [inline src/server/IoDispatchers.cpp:182]: I noticed a slight "bug" (I don't think it can occur tbh) in my first commit and added a `Globals.handoffConsoleClsid.reset();` below in my last commit.
While doing so I noticed that the order of if checks in `_shouldAttemptHandoff` is inoptimal: It first checked `_isInteractiveUserSession` instead of checking cheap boolean flags first, so I changed this code. Let me know if I should revert it for a more compact PR.

File: src/host/srvinit.cpp (hunk 1 above)
Comment (zadjii-msft) [inline src/host/srvinit.cpp:444]: interesting, so we're being clever here - we've been invoked for "default console", but we didn't find a delegation terminal, so we can assume that Terminal Stable should be the default terminal, cause the only way to get here is if we're the OpenConsole for a DxD invocation. Makes sense.

File: src/server/IoDispatchers.cpp [hunk 3]
```diff
@@ -269,6 +269,118 @@ static bool _shouldAttemptHandoff(const Globals& globals,
...
+        if (Globals.defaultTerminalMarkerCheckRequired)
+        {
+            ::Microsoft::WRL::ComPtr<IDefaultTerminalMarker> marker;
+            if (FAILED(handoff.As(&marker)))
+            {
+                Globals.handoffConsoleClsid.reset();
+                Globals.defaultTerminalMarkerCheckRequired = false;
+                return;
+            }
+        }
```
Comment (zadjii-msft) [inline src/server/IoDispatchers.cpp:297]: note: this is the new blob of code here

File: src/propslib/DelegationConfig.cpp [hunk 4]
```diff
@@ -238,7 +245,10 @@ try
     {
         if (pkg.console.clsid == defCon && pkg.terminal.clsid == defTerm)
         {
-            chosenPackage = pkg;
+            if (!isTerminalAsDefault || pkg.console.allowDefaultTerminalHandoff)
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/propslib/DelegationConfig.cpp:248]: Ohhhhhhhhhhhhhhhhhhhhh okay I see now. The `CheckShouldTerminalBeDefault` blocks will set the respective `defCon` and `defTerm` to the clsid's for Terminal Stable. Okay. Now it all makes sense.

**Review by zadjii-msft — APPROVED** [review]
Okay, this all looks like it'll work Terminal-side, and conhost side. My only recommendation would be add more comments. A lot of this defterm stuff is hard to parse, so more comments never hurt. 

The code is good enough though, so :shipit:

File: src/propslib/DelegationConfig.cpp [hunk 5]
```diff
@@ -127,6 +127,12 @@ HRESULT _lookupCatalog(PCWSTR extensionName, std::vector<T>& vec) noexcept
         ComPtr<IMap<HSTRING, IInspectable*>> map;
         RETURN_IF_FAILED(properties.As(&map));
 
+        {
+            boolean allowDefaultTerminalHandoff;
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/propslib/DelegationConfig.cpp:131]: Yeah it's weird. I was super surprised myself. 🤨

File: src/propslib/DelegationConfig.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (lhecker) [reply src/propslib/DelegationConfig.cpp:248]: I'm not quite sure how to best structure the code here. But it would probably flow better in general if `s_GetDefaultConsoleId` and `s_GetDefaultTerminalId` were one function. 🤔

File: src/server/IoDispatchers.cpp (hunk 3 above)
Comment (lhecker) [reply src/server/IoDispatchers.cpp:297]: Yep the remaining function is unchanged. It was extracted into a standalone function so that I can write `return;` within this new if condition.

File: src/propslib/DelegationConfig.cpp [hunk 6]
```diff
@@ -225,7 +231,8 @@ try
     // We also find the default here while we have the list of available ones so
     // we can return the opaque structure instead of the raw IID.
     IID defCon;
-    LOG_IF_FAILED(s_GetDefaultConsoleId(defCon));
+    bool isTerminalAsDefault;
```
Comment (carlos-zamora) [inline src/propslib/DelegationConfig.cpp:234]: `isTerminalAsDefault` is false here, right?

**Review by carlos-zamora — DISMISSED** [review]
Looks good to me. Thanks!

**Comment by carlos-zamora** [comment]
Dustin wanted to take a look at this, so I'm leaving it unmerged for now :)

File: src/propslib/DelegationConfig.cpp (hunk 6 above)
Comment (lhecker) [reply src/propslib/DelegationConfig.cpp:234]: No, by default primitive types have an undefined value before they're initialized.
But `s_GetDefaultConsoleId` will always set this variable no matter what.

**Comment by carlos-zamora** [comment]
@msftbot make sure @DHowett signs off on this

File: src/propslib/DelegationConfig.cpp [hunk 7]
```diff
@@ -293,10 +305,10 @@ CATCH_RETURN()
 
     auto hr = s_Get(DELEGATION_TERMINAL_KEY_NAME, iid);
 
-    auto defApp = false;
-    if (SUCCEEDED(Microsoft::Console::Internal::DefaultApp::CheckShouldTerminalBeDefault(defApp)) && defApp)
+    if (FAILED(hr))
     {
-        if (FAILED(hr))
+        auto defApp = false;
+        if (SUCCEEDED(Microsoft::Console::Internal::DefaultApp::CheckShouldTerminalBeDefault(defApp)) && defApp)
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/propslib/DelegationConfig.cpp:311]: Yeah, it should compile out the entire thing in OpenConsole. 🤔

**Comment by lhecker** [comment]
I've rewritten this PR, but haven't tested it yet and so converted this PR to a Draft.
I think it should work however, so please feel free to review if you have time to spare. 🙂

**Comment by lhecker** [comment]
This is how the new option looks like:
image

File: src/cascadia/TerminalSettingsModel/CascadiaSettings.cpp [hunk 8]
```diff
@@ -1120,12 +1121,27 @@ void CascadiaSettings::CurrentDefaultTerminal(const Model::DefaultTerminal& term
...
+        const auto onexit = wil::scope_exit([&]() {
+            latch.count_down();
+        });
+        co_await winrt::resume_background();
+        result = DefaultTerminal::Available();
+    }();
+
+    latch.wait();
+    _defaultTerminals = winrt::single_threaded_observable_vector(std::move(result.first));
+    _currentDefaultTerminal = std::move(result.second);
```
Comment (lhecker) [inline src/cascadia/TerminalSettingsModel/CascadiaSettings.cpp:1144]: In hindsight this is a clear issue, as `DefaultTerminal::Available` accesses the app catalog, but for some inexplicable reason I only started hitting this after I made the changes in this PR. And once they started I hit them _all_ the time. I couldn't complete testing this PR without these changes.

File: src/propslib/DelegationConfig.cpp [hunk 9]
```diff
@@ -261,90 +223,62 @@ CATCH_RETURN()
...
-    DWORD bytesNeeded = 0;
-    auto result = RegistrySerialization::s_QueryValue(startupKey.get(),
-                                                      value,
-                                                      0,
-                                                      REG_SZ,
-                                                      nullptr,
-                                                      &bytesNeeded);
-
-    if (NTSTATUS_FROM_WIN32(ERROR_SUCCESS) != result)
+    if (values[0] == CLSID_Default || values[1] == CLSID_Default)
```
Comment (zadjii-msft) [inline src/propslib/DelegationConfig.cpp:273]: notes: if either was missing or failed to parse, use {default, 0, 0}

File: src/propslib/DelegationConfig.cpp [hunk 10]
```diff
@@ -261,90 +223,62 @@ CATCH_RETURN()
...
-                                                                  value,
-                                                                  bytesNeeded,
-                                                                  REG_SZ,
-                                                                  reinterpret_cast<BYTE*>(buffer.get()),
-                                                                  &bytesUsed));
-
-    RETURN_IF_FAILED(IIDFromString(buffer.get(), &iid));
-
-    return S_OK;
+    if (values[0] == CLSID_Conhost || values[1] == CLSID_Conhost)
```
Comment (zadjii-msft) [inline src/propslib/DelegationConfig.cpp:277]: Was the saved value SPECIFICALLY conhost? then return that



oh ho ho so I see, this is where we're being tricky now. Now conhost isn't {0}, it's got it's own GUID now. That makes sense.

File: src/propslib/DelegationConfig.cpp [hunk 11]
```diff
@@ -222,29 +194,19 @@ try
...
-    LOG_IF_FAILED(s_GetDefaultConsoleId(defCon));
-    IID defTerm;
-    LOG_IF_FAILED(s_GetDefaultTerminalId(defTerm));
-
-    // The default one is the 0th one because that's supposed to be the inbox conhost one.
-    auto chosenPackage = packages.at(0);
-
-    // Search through and find a package that matches. If we failed to match because
-    // it's torn across multiple or something not in the catalog, we'll offer the inbox conhost one.
+    // Search through and find a package that matches.
```
Comment (zadjii-msft) [inline src/propslib/DelegationConfig.cpp:197]:
```suggestion
    // Get the currently set default console/terminal. 
    // Then, search through and find a package that matches. If we find one, then return it. 
```

File: src/propslib/DelegationConfig.cpp [hunk 12]
```diff
@@ -29,27 +29,11 @@ using namespace ABI::Windows::ApplicationModel::AppExtensions;
...
-
-    T useInbox = { 0 };
-    useInbox.clsid = { 0 };
-    // CLSID of 0 will be sentinel to say "inbox console" or something.
-    // The UI displaying this information will have to go look up appropriate strings
-    // to convey that message.
-
-    vec.push_back(useInbox);
-
-    auto coinit = wil::CoInitializeEx(COINIT_APARTMENTTHREADED);
```
Comment (zadjii-msft) [inline src/propslib/DelegationConfig.cpp:51]: We pretty sure the propsheet initializes COM itself? Note that the propsheet can get loaded outside of the context of conhost, just by right-clicking on a `.lnk` to a commandline application.

**Review by zadjii-msft — APPROVED** [review]
Okay, I read over this all, it makes sense to me. I again left notes as I went - my nit again is that all this defterm stuff is already hard to parse so more comments is always better.

File: src/cascadia/TerminalSettingsEditor/Launch.xaml [hunk 13]
```diff
@@ -93,7 +93,7 @@
 
                                     <Grid.RowDefinitions>
                                         <!--  profile name  -->
-                                        <RowDefinition Height="*" />
+                                        <RowDefinition Height="20" />
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/cascadia/TerminalSettingsEditor/Launch.xaml:96]: Unfortunately we only get the layout in the screenshot above, by specifying the height explicitly twice. The reason is that the flyout (and only the flyout) seemingly doesn't respect the visibility state of the two `TextBlock`s. Due to that the dropdown box will have the correct height (1 line high) but the flyout list will be 2 lines high, even with both `TextBlock`s collapsed. I'm not entirely sure how such a bug can occur, but it does. 😄

File: src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp [hunk 14]
```diff
@@ -61,9 +61,11 @@ static auto extractValueFromTaskWithoutMainThreadAwait(TTask&& task) -> decltype
     til::latch latch{ 1 };
 
     const auto _ = [&]() -> winrt::fire_and_forget {
+        const auto cleanup = wil::scope_exit([&]() {
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp:64]: Yeah I realized this mistake when I wrote the code for `CascadiaSettings.cpp` and wanted to fix `extractValueFromTaskWithoutMainThreadAwait` as well.

File: src/inc/conint.h [hunk 15]
```diff
@@ -47,7 +47,7 @@ namespace Microsoft::Console::Internal
 
     namespace DefaultApp
     {
-        [[nodiscard]] HRESULT CheckDefaultAppPolicy(bool& isEnabled) noexcept;
-        [[nodiscard]] HRESULT CheckShouldTerminalBeDefault(bool& isEnabled) noexcept;
+        [[nodiscard]] bool CheckDefaultAppPolicy() noexcept;
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/inc/conint.h:50]: Yep! I felt like this made writing the code more comfortable and noticed that the internal code never returns anything but `S_OK` anyways.

File: src/propslib/DelegationConfig.hpp [hunk 16]
```diff
@@ -18,83 +18,104 @@ Author(s):
...
     };
 
     [[nodiscard]] static HRESULT s_GetAvailablePackages(std::vector<DelegationPackage>& packages, DelegationPackage& def) noexcept;
 
     [[nodiscard]] static HRESULT s_SetDefaultByPackage(const DelegationPackage& pkg, const bool useRegExe = false) noexcept;
 
-    [[nodiscard]] static HRESULT s_GetDefaultConsoleId(IID& iid) noexcept;
-    [[nodiscard]] static HRESULT s_GetDefaultTerminalId(IID& iid) noexcept;
+    static constexpr CLSID CLSID_Default{};
+    static constexpr CLSID CLSID_Conhost{ 0xb23d10c0, 0xe52e, 0x411e, { 0x9d, 0x5b, 0xc0, 0x9f, 0xdf, 0x70, 0x9c, 0x7d } };
```
Comment (DHowett) [inline src/propslib/DelegationConfig.hpp:105]: What happens when a user installs Terminal 1.15 on a Windows 11 machine that doesn't know about the new guid that means "conhost"?

File: src/server/IoDispatchers.cpp [hunk 17]
```diff
@@ -269,6 +269,118 @@ static bool _shouldAttemptHandoff(const Globals& globals,
...
+        // Get the class/interface to the handoff handler. Local machine only.
+        ::Microsoft::WRL::ComPtr<IConsoleHandoff> handoff;
+        THROW_IF_FAILED(CoCreateInstance(Globals.delegationPair.console, nullptr, CLSCTX_LOCAL_SERVER, IID_PPV_ARGS(&handoff)));
+
+        if (Globals.defaultTerminalMarkerCheckRequired)
+        {
+            ::Microsoft::WRL::ComPtr<IDefaultTerminalMarker> marker;
+            if (FAILED(handoff.As(&marker)))
+            {
+                Globals.delegationPair = DelegationConfig::ConhostDelegationPair;
```
Comment (DHowett) [inline src/server/IoDispatchers.cpp:293]: Will this result in the propsheet showing "Windows Console Host"? Or will it show "Let Windows Decide" still?

File: src/propslib/DelegationConfig.hpp (hunk 16 above)
Comment (lhecker) [reply src/propslib/DelegationConfig.hpp:105]: We can't use velocity, so I'd solve this with a version check in WT potentially?

File: src/server/IoDispatchers.cpp (hunk 17 above)
Comment (DHowett) [reply src/server/IoDispatchers.cpp:293]: oh, it will still show LWD because this code isn't used by propsheet

Comment (lhecker) [reply src/server/IoDispatchers.cpp:293]: The propsheet calls `s_GetAvailablePackages` itself again and doesn't depend on `Globals`.

File: src/propslib/DelegationConfig.cpp (hunk 12 above)
Comment (lhecker) [reply src/propslib/DelegationConfig.cpp:51]: I just moved the initialization one call up so that we don't have to do this twice.

**Review by DHowett — APPROVED** [review]
